SmartEVSE-3/src/evse.cpp

```diff
@@ -1488,21 +1494,46 @@ void printStatus(void)
 }
 
 // Recompute State of Charge, in case we have a known initial state of charge
-// This function is called by kWh logic
-void RecomputeSoC(void){
-    if (InitialSoC > 0)
-        ComputedSoC = InitialSoC + (EnergyCharged / 280);
+// This function is called by kWh logic and after an EV state update through API, Serial or MQTT
+void RecomputeSoC(void) {
+    if (FullSoC > 0 && EnergyCapacity > 0) {
+        if (InitialSoC == FullSoC) {
+            // We're already at full SoC
+            ComputedSoC = FullSoC;
+            RemainingSoC = 0;
+        } else if (EnergyRequest > 0) {
+            // Attempt to use EnergyRequest to determine SoC with greater accuracy
+            // We're adding 50 Wh to EnergyCharged here to be sure we reach 100% ComputedSoC and compensate for losses
+            uint32_t RemainingEnergyWh = (EnergyCharged > 0 ? EnergyRequest - (EnergyCharged + 50) : EnergyRequest);
+            if (RemainingEnergyWh > 0) {
+                ComputedSoC = FullSoC - (((double) RemainingEnergyWh / EnergyCapacity) * 100);
+                RemainingSoC = FullSoC - ComputedSoC;
+                return;
+            } else {
+                ComputedSoC = FullSoC;
+                RemainingSoC = 0;
+            }
+        } else if (InitialSoC > 0) {
+            // Fall back to rough estimate based on InitialSoC if we do not know the requested energy
+            ComputedSoC = InitialSoC + (((double) EnergyCharged / EnergyCapacity) * 100);
+            RemainingSoC = FullSoC - ComputedSoC;
+        }
+    }
+    // There's also the possibility an external API/app is used for SoC info. In such case, we allow setting ComputedSoC directly.
 }
 
 // EV disconnected from charger. Triggered after 60 seconds of disconnect
 // This is done so we can "re-plug" the car in the Modem process without triggering disconnect events
 void DisconnectEvent(void){
     _LOG_A("EV disconnected for a while. Resetting SoC states");
     ModemStage = 0; // Enable Modem states again
+    LeaveModemDeniedStateTimer = 0; // Reset ModemDeniedStateTimer
```

Comment:
This is the only timer which is reset in this function

Reply:
Yeah it may be redundant as I don't think it actually does anything. I will remove it.